### PR TITLE
CFE-2263 Fix select_end EOF matching.

### PR DESCRIPTION
**Pull request by pasinskim**
Make sure that if select_end is not matched in file, EOF is used as
and of region by default.
To change default behavior, 'select_end_match_eof' was introduced.
The default value is true, so EOF is threated as end of region by default.

Changelog: EOF is implicitly matched as an end of the region in edit_line
promises. To change default behavior 'select_end_match_eof' parameter
was introduced. (CFE-2263)

### CODE REVIEW

**Comment by pasinskim** [comment]
@kacf @jimis can you take a look at this one?

File: libpromises/mod_files.c [hunk 1]
```diff
@@ -71,6 +71,7 @@ static const ConstraintSyntax select_region_constraints[] =
     ConstraintSyntaxNewBool("include_end_delimiter", "Whether to include the section delimiter. Default value: false", SYNTAX_STATUS_NORMAL),
     ConstraintSyntaxNewString("select_start", CF_ANYSTRING, "Regular expression matching start of edit region", SYNTAX_STATUS_NORMAL),
     ConstraintSyntaxNewString("select_end", CF_ANYSTRING, "Regular expression matches end of edit region from start", SYNTAX_STATUS_NORMAL),
+    ConstraintSyntaxNewBool("select_end_match_eof", "Whenever to include EOF as end of the region. Default value: true", SYNTAX_STATUS_NORMAL),
```
Comment (jimis) [inline libpromises/mod_files.c]: ```s/Whenever/Whether/```

**Comment by jimis** [comment]
Overall looks good, without going into the edit_lines code logic.

File: libpromises/attributes.c [hunk 2]
```diff
@@ -1489,6 +1489,17 @@ EditRegion GetRegionConstraints(const EvalContext *ctx, const Promise *pp)
...
     e.include_end = PromiseGetConstraintAsBoolean(ctx, "include_end_delimiter", pp);
+    char *value = PromiseGetConstraintAsRval(pp, "select_end_match_eof", RVAL_TYPE_SCALAR);
+    if (value && (strcmp(value, "false") == 0))
+    {
+        e.select_end_match_eof = false;
+    }
+    else
+    {
+        /* Default value for 'select_end_match_eof' is true. */
+        e.select_end_match_eof = true;
```
Comment (kacf) [inline libpromises/attributes.c]: It seems like the consensus in [CFE-2263](https://tracker.mender.io/browse/CFE-2263) is to have this be false by default.

File: libpromises/mod_files.c (hunk 1 above)
Comment (kacf) [inline libpromises/mod_files.c]: And default value should be false.

File: tests/acceptance/10_files/07_delete_lines/no_matching_end.cf [hunk 3]
```diff
@@ -73,7 +67,8 @@ bundle edit_line test_delete(str)
 body select_region test_select
 {
       select_start => "BEGIN";
-      select_end => "END"; # anchored regex doesn't match due to whitespace
+      select_end => "END"; # anchored regex doesn't match due to whitespace, but 
+                           # 'select_end_match_eof' is true by default.
```
Comment (kacf) [inline tests/acceptance/10_files/07_delete_lines/no_matching_end.cf]: This test will have to be changed if the default changes, I assume.

File: tests/acceptance/10_files/07_delete_lines/select_end_match_eof_false.cf [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  delete_lines:
+      "$(str)"
+      select_region => test_select;
+}
+
+body select_region test_select
+{
+      select_start => "BEGIN";
+      select_end => "END"; # anchored regex doesn't match due to whitespace, but 
+                           # 'select_end_match_eof' is true by default.
```
Comment (kacf) [inline tests/acceptance/10_files/07_delete_lines/select_end_match_eof_false.cf]: Will also have to change if default changes.

**Comment by kacf** [comment]
If you change the default, remember to change the changelog message as well.

**Comment by kacf** [comment]
Looks good otherwise!

File: libpromises/mod_files.c (hunk 1 above)
Comment (pasinskim) [inline libpromises/mod_files.c]: From today on... fixed!

File: libpromises/attributes.c (hunk 2 above)
Comment (pasinskim) [inline libpromises/attributes.c]: Fixing!

File: cf-agent/files_editline.c [hunk 5]
```diff
@@ -683,6 +683,15 @@ static PromiseResult VerifyLineInsertions(EvalContext *ctx, const Promise *pp, E
         result = PromiseResultUpdate(result, PROMISE_RESULT_INTERRUPTED);
         return result;
     }
+    
+    if (!end_ptr && a.region.select_end && !a.region.select_end_match_eof)
+    {
+        cfPS(ctx, LOG_LEVEL_VERBOSE, PROMISE_RESULT_INTERRUPTED, pp, a,
+            "The promised end pattern '%s' was not found when selecting region to insert in '%s'",
```
Comment (nickanderson) [inline cf-agent/files_editline.c]: nice and clear message

**Comment by kacf** [comment]
A couple of typos in the commit messages (particularly the changelog):

* EOF is  matched as an end of the region in edit_line promises is 'select_end_match_eof' parameter is true.
 * EOF is matched as an end of the region in edit_line promises **only if** 'select_end_match_eof' parameter is true.
* instert_lines is no longer implicitly matching EOF as end of the region if 'select_end' pattern is not matched
 * **insert_lines** is no longer implicitly matching EOF as end of the region if 'select_end' pattern is not matched

Otherwise looks good!

**Comment by pasinskim** [comment]
@kacf thanks. Commit messages are fixed now.

**Comment by kacf** [comment]
Changelog: "**only if**", not "only is" :-)

**Comment by pasinskim** [comment]
@kacf: fixed now. Eystein and Nick confirmed that we can merge both commits.

**Comment by kacf** [comment]
Looks good!

**Comment by kacf** [comment]
trigger build

**Comment by pasinskim** [comment]
Looks like build is green, isn't it?

**Comment by kacf** [comment]
Yep, it's green!

**Pull request merged**